### PR TITLE
[locator] adding Get<Type> to InstanceLocator

### PR DESCRIPTION
**Pull request by abtink**
This commit changes how the objects in OpenThread access each other.
It adds a template `Get<Type>()` method in `InstanceLocator`. This
method returns a reference to a given `Type` object belonging to the
OpenThread instance (e.g. `Get<MeshForwarder>()` returns a reference
to `MeshForwarder` object on the OpenThread instance). The
`InstanceLocator` is used as base class of all OpenThread classes so
every class can easily access any other object. This commit also
changes how the main instance is retrieved in `InstanceLocator` for
the single-instance case. The method `GetInstance()` directly uses the
raw buffer `gInstanceRaw`. This change helps make the `GetInstance()`
and in turn all `Get<Type>()` methods in-line. This commit also
removes the existing getters across all classes to use the new
`Get<Type>()` model.

------------------------------------

This change provides the following advantages:
- Simplifies the code and makes it easier to read by removing long chains of getters calls:
```
GetNetif().GetKeyManager()                               --> Get<KeyManager>()
GetNetif().GetIp6().GetMpl()                             --> Get<Ip6::Mpl>()
GetNetif().GetMac().GetFilter()                          --> Get<Mac::Filter>()
GetNetif().GetNetworkDataLeader()                        --> Get<NetworkData::Leader>()
GetNetif().GetMle().GetChildTable()                      --> Get<ChildTable>()
GetNetif().GetMeshForwarder().GetDataPollManager()       --> Get<DataPollManager>()
GetNetif().GetMeshForwarder().GetSourceMatchController() --> Get<SourceMatchController>()
```
- Removes a lot of simple getter methods in different classes (all getters are now in `instance.hpp`). Overall it helps reduce source code size:
```
 121 files changed, 1848 insertions(+), 2505 deletions(-)
```

- Reduces code size a bit (the reduction is mainly coming from `Get<Type>()` becoming fully in-lined, mapping to a constant pointer). No impact on RAM.
```
 +arm-none-eabi-size output/cc2538/bin/ot-cli-ftd
   text	   data	    bss	    dec	    hex	filename
 207000	   5252	  22344	 234596	  39464	output/cc2538/bin/ot-cli-ftd  (this branch)
 208468	   5252	  22344	 236064	  39a20	output/cc2538/bin/ot-cli-ftd  (master)
 (-1468)   

+arm-none-eabi-size output/cc2538/bin/ot-cli-mtd
   text	   data	    bss	    dec	    hex	filename
 150036	   5252	  18440	 173728	  2a6a0	output/cc2538/bin/ot-cli-mtd (this branch)
 150476	   5252	  18440	 174168	  2a858	output/cc2538/bin/ot-cli-mtd (master)
 (-440)

+arm-none-eabi-size output/cc2538/bin/ot-ncp-ftd
   text	   data	    bss	    dec	    hex	filename
 210436	   5252	  22840	 238528	  3a3c0	output/cc2538/bin/ot-ncp-ftd (this branch)
 211920	   5252	  22840	 240012	  3a98c	output/cc2538/bin/ot-ncp-ftd (master)
 (-1484)

+arm-none-eabi-size output/cc2538/bin/ot-ncp-mtd
   text	   data	    bss	    dec	    hex	filename
 156900	   5252	  18936	 181088	  2c360	output/cc2538/bin/ot-ncp-mtd (this branch)
 157364	   5252	  18936	 181552	  2c530	output/cc2538/bin/ot-ncp-mtd (master)
 (-464)
```
- Makes it easier to later move around objects in the `Instance` hierarchy (if needed).

### CODE REVIEW

File: src/core/meshcop/dataset_local.hpp [hunk 1]
```diff
@@ -143,6 +143,13 @@ class DatasetLocal : public InstanceLocator
     int Compare(const Timestamp *aCompare);
 
 private:
+    /**
+     * This method re-defines template `Get<Type>()` as the `InstanceLoctaor` (base class) definition is shadowed by
```
[the remark this answers is not in the document]
Comment (bukepo) [reply src/core/meshcop/dataset_local.hpp]: Shall we rename the existing `Get()` to `Read()` to better reflect the functionality and avoid conflicts in this PR or another PR?

[the remark this answers is not in the document]
Comment (abtink) [reply src/core/meshcop/dataset_local.hpp]: :+1:  Good idea. 
I was considering the same but did not go for it as this PR is already rather big.
Let me see if I can rename this in a separate commit.

[the remark this answers is not in the document]
Comment (abtink) [reply src/core/meshcop/dataset_local.hpp]: Submitted https://github.com/openthread/openthread/pull/3723 (as a separate PR as well) to make it easier to review.

**Comment by abtink** [comment]
Rebased this PR on master with https://github.com/openthread/openthread/pull/3723 merged.

**Pull request merged**